dotenv_codegen_implementation/src/lib.rs

```diff
@@ -7,41 +7,50 @@ use proc_macro_hack::proc_macro_hack;
 use quote::quote;
 use syn::parse::Parser;
 use syn::punctuated::Punctuated;
+use syn::spanned::Spanned;
 use syn::Token;
 
 #[proc_macro_hack]
 pub fn dotenv(input: TokenStream) -> TokenStream {
     if let Err(err) = dotenv::dotenv() {
-        panic!("Error loading .env file: {}", err);
+        let msg = format!("Error loading .env file: {}", err);
+        return quote! {
+            compile_error!(#msg);
+        }
+        .into();
     }
 
-    // Either everything was fine, or we didn't find an .env file (which we ignore)
-    expand_env(input)
+    match expand_env(input) {
+        Ok(stream) => stream,
+        Err(e) => e.to_compile_error().into(),
+    }
 }
 
-fn expand_env(input_raw: TokenStream) -> TokenStream {
-    let args = <Punctuated<syn::LitStr, Token![,]>>::parse_terminated
-        .parse(input_raw)
-        .expect("expected macro to be called with a comma-separated list of string literals");
+fn expand_env(input_raw: TokenStream) -> syn::Result<TokenStream> {
+    let args = <Punctuated<syn::LitStr, Token![,]>>::parse_terminated.parse(input_raw)?;
 
     let mut iter = args.iter();
 
-    let var_name = match iter.next() {
-        Some(s) => s.value(),
-        None => panic!("expected 1 or 2 arguments, found none"),
-    };
-
-    let err_msg = match iter.next() {
-        Some(lit) => lit.value(),
-        None => format!("environment variable `{}` not defined", var_name),
-    };
-
+    let var_name = iter
+        .next()
+        .ok_or_else(|| syn::Error::new(args.span(), "expected at least 1 argument"))?;
+    let err_msg = iter.next();
     if iter.next().is_some() {
-        panic!("expected 1 or 2 arguments, found 3 or more");
+        return Err(syn::Error::new(
+            args.span(),
+            "expected a maximum of 2 arguments",
+        ));
     }
 
-    match env::var(var_name) {
-        Ok(val) => quote!(#val).into(),
-        Err(VarError::NotPresent) | Err(VarError::NotUnicode(_)) => panic!("{}", err_msg),
+    match env::var(var_name.value()) {
+        Ok(val) => Ok(quote!(#val).into()),
+        Err(VarError::NotPresent) | Err(VarError::NotUnicode(_)) => Err(syn::Error::new(
```

Comment:
I think we should break up this match arm now that we're producing more descriptive error messages. `NotUnicode` doesn't match with `environment variable {} not defined`

Reply:
Taking a second look at this, I want to mirror the behavior of [`std::env!`](https://github.com/rust-lang/rust/blob/1033351a51dd3ca342a83d4be13f7554f0b4fb1e/src/librustc_builtin_macros/env.rs#L51-L93), which actually doesn't differentiate between the two errors. Not that that means we can't though